### ISSUE TITLE
Enhancements - grid viewport classes and visibility utility classes

### ISSUE THREAD

**Issue by ncksllvn** [uswds/uswds#2173]
I'm a developer on the team for vets.gov and if possible we'd love some insight into what the road ahead looks like for USWDS. Our project was initially started leveraging the Foundation CSS framework and although we've begun migrating to USWDS, there are certain features of Foundation that we can't do without and now we're deciding how to proceed. 

The main features of Foundation that we're invested in and that are not already implemented in USWDS are the visibility utility classes and control over the grid for certain viewports (i.e. `usa-medium-width-one-half` or something like that.) Are there any plans for these to be included in the project, and if so are you open to working together to implement them? If not, we'll just backfill them in our own project as we remove Foundation, but we thought it could be a useful collaboration if these are planned features for USWDS and it would really help us out.

Note - I sent an email last week about this but didn't receive a response.

**Comment by maya** [comment]
Hey Nick, thanks for opening this issue. Just realized you didn't get my email so re-posting here as well.

That's exciting to hear about how you're using the Standards. We've heard of folks wanting more viewport control in the past like you mentioned so we'll likely take a look at it in the future. I don't know when we'll be working on rethinking grids, but can check with the core team and see how it fits in our roadmap and if/when would be good time to collaborate. In the meantime, feel free to go ahead with your modification and let us know how it progresses.

**Comment by ncksllvn** [comment]
Thanks for your response Maya! Okay, we'll figure something out for our project and check back periodically. Great job with the project so far by the way!

**Comment by emes001** [comment]
Fellow dev over at @USOPM here. I second more granular control of the grid system, based on a specific viewport. In lieu of this missing feature, I'm relying more on custom classes at the moment (more-so because this is for a prototype). Do you have a timeline for 2.0 and/or is this currently being worked on?

**Comment by stphnwlkr** [comment]
A third. I have started looking at alternatives as well because the grid is too limiting. The CMS team has done a nice job as well as the more popular frameworks.

**Comment by maya** [comment]
@emes001 the timeline for 2.0 is likely in the next few months, in which updates to the grid would be included. The main focus right now is on our utility classes feature and will likely be taking a closer look at grids soon after that feature is complete.

**Comment by annekainicUSDS** [comment]
Our team has done some thinking about the specific requirements for any enhancements to the grid that we would benefit from, and how those might be implemented. We thought we would share them here to get your thoughts and feedback!

#### Desired Enhancements:
- Mobile-First Grid System
- Responsive Utilities

Follow-up to https://github.com/uswds/uswds/issues/2173.

# Mobile-First Grid System

## Current Limitations
The current 12-column grid system works very well on large screens, but is limited when developing layouts that require more granular control for smaller devices.

For example, the standard layout for content on [Vets.gov](https://www.vets.gov/) consists of a sidebar and main content. The sizes were defined as one fourth for the sidebar and three fourths for the main content. This works well for large screens. However, because on medium screens the grid is only six columns, the layout is broken - the sidebar remains the same size while the three-fourths content expands to 100%. To fix, we had to apply a custom override so that the medium screen size will be ~75%.

```html
<!--
// The following HTML layout required this CSS to keep the main content from collapsing below the sidebar:
@media screen and (min-width: 768px)
.usa-grid .usa-width-three-fourths {
width: 74.41059%;
}
-->
<div class="usa-grid">
<div id="sidebar" class="usa-width-one-fourth"></div>
<div id="content" class="usa-width-three-fourths"></div>
</div>
```

<img width="1391" alt="screen shot 2018-03-23 at 1 44 10 pm" src="https://user-images.githubusercontent.com/25183456/37848014-13300762-2ea9-11e8-8aa7-e86ad93acc84.png">


Another example illustrates the need for small screen grid styling. Our current layout for forms includes horizontally aligned navigation buttons at the bottom of each form page. These buttons are horizontally aligned even at the smallest screen widths (see screenshot below). We currently use Foundation classes to manage this alignment ([see here](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/js/common/schemaform/containers/FormPage.jsx#L123)).

<img width="363" alt="screen shot 2018-03-23 at 10 51 48 am" src="https://user-images.githubusercontent.com/25183456/37848032-1d8b4e42-2ea9-11e8-8776-037a727d00c2.png">


## Desired Enhancement
Classes that allowed an explicit screen size would be ideal. The following code was adapted from other grid systems, such as Bootstrap. When the screen size is omitted, it assumes all screens.

```scss
.usa-width-one-fourth // all screens, including small-width screen sizes
.usa-width-medium-one-fourth // screens medium and up
.usa-width-large-one-fourth // screens large and up (in case there were an extra-large or something)
```
Using the above, we could define the sidebar/content layout as the following without any CSS overrides:
```html
<div class="usa-grid">
<div id="sidebar" class="usa-width-medium-one-fourth"></div>
<div id="content" class="usa-width-medium-three-fourths"></div>
</div>
```

## Backwards-Compatibility
Because some websites may be reliant on the automatic responsive behavior of the current grid system, an additional grid class may be required to maintain backwards-compatibility. For example, we could use a wrapping element, `<div class="usa-grid-manual">`, for the `usa-width-*screen*-one-fourth` class to prevent conflicts with the old grid system and preserve backwards compatibility for the websites that work with the current grid system. It would look like:

```html
<div class="usa-grid-manual"><div class="usa-width-large-one-half usa-width-medium-one-whole"></div></div>
```

# Visiblity Classes
Currently there is no method of toggling the visiblity of an element based on the screen size. Based on Foundation's, these could be implemented as:

```scss
.usa-hide // hidden on all screens
.usa-hide-small  // hidden on small and up
.usa-hide-medium // hidden on medium and up
.usa-hide-large // hidden on large and up

.usa-hide-small-only  // hidden only on small
.usa-hide-medium-only // hidden only on medium
.usa-hide-large-only // hidden only on large

.usa-show-medium // show on medium and up
.usa-show-large // show on large and up

.usa-show-small-only  // show only on small
.usa-show-medium-only // show only on medium
.usa-show-large-only // show only on large
```

For example, in the layout above, we wrote CSS to hide the sidebar on small screens and reveal a button for sliding in the sidebar as a drawer. With utility classes, we could have written following:

```html
<button class="usa-button usa-show-small-only">Toggle Drawer</button>
<div class="usa-grid">
<div id="sidebar" class="usa-hidden-small-only usa-width-medium-one-fourth"></div>
<div id="content" class="usa-width-medium-three-fourths"></div>
</div>
```

**Comment by maya** [comment]
This is really useful feedback @annekainicUSDS, thanks! Appreciate the thought behind backwards compatibility too. Also, a former team member had started some thinking around this too: #2137.

**Comment by ncksllvn** [comment]
Hi @maya! Based on @annekainicUSDS's suggestions would your team be open to contributions from the team at the Dept of Veteran Affairs? We're looking to do some revamping in the coming weeks and would love to work with you on this.

**Comment by ncksllvn** [comment]
@emes001 @stphnwlkr - would the suggestions above work for your use case as well?

**Comment by maya** [comment]
@ncksllvn yes, we'd be glad to collaborate on this. We're still in the middle of our work on utilities, but we'll touch base again in a few weeks. We'll be taking grids on next following utilities.

**Comment by maya** [comment]
Hi all, we'd like to share some updates on this. We're working on a proposal for a brand new grid system which you can take a look at here which has the viewport visibility toggles: https://federalist-proxy.app.cloud.gov/preview/gsa/10x/dw-design-refinements/layout-grid/

Notes: 
- This would be a breaking change as it changes the names of the classes.
- The idea is you'd add the viewport to the columns: `.tablet:g-col-4` 
  - Could use `.sm:g-col-4` rather than tablet bc tablet size range is large and subjective

Thoughts? Feel free to post general thoughts here and specific thoughts on the code in the PR #2422.

/cc @ncksllvn @emes001 @stphnwlkr @annekainicUSDS

**Comment by maya** [comment]
The new grid is implemented in #2422. The visibility classes are also implemented in the `release-2.0` branch.

Will be available in the next 2.0 release. Thanks for your thoughts, everyone!
